clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchBuffer.java

```diff
@@ -35,19 +41,24 @@
  *
  * <p/>
  *
- * <em>Note</em>: this class is not thread-safe and is intended to only be used from a single thread.
+ * <em>Note</em>: this class is thread-safe with the intention that {@link CompletedFetch the data} will be
```

Comment:
Currently, `fetchBuffer.setNextInLineFetch` and `fetchBuffer.poll` are separate operations and we expect the caller to call them in the right order to avoid a partition missing in FetchBuffer in the transition phase. It still leaves us with the situation that a partition could be in both completedFetches and nextInLineFetch at a particular time. It's not a problem for now, but it may be in the future.  Could we make them an atomic operation? If not, could we add a comment to document the correct usage of the api and the impact on partition being duplicated in completedFetches and nextInLineFetch?

Reply:
OK, I'll look at this again.

Reply:
I've filed KAFKA-15640 to look into this in more detail.